### PR TITLE
Adding support for Telemetry Keys on C# SDK

### PR DESCRIPTION
**Pull request by rgomez391**
Adding support for Telemetry Key on C# SDK by creating dedicated templates for Events API.

### CODE REVIEW

File: targets/csharp/templates/PlayFab_Events_InstanceAPI.cs.ejs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+                PlayFabSettings.GlobalErrorHandler?.Invoke(error);
+                return new PlayFabResult<<%- apiCall.result %>> { Error = error, CustomData = customData };
+            }
+
+            var resultRawJson = (string)httpResult;
+            var resultData = PluginManager.GetPlugin<ISerializerPlugin>(PluginContract.PlayFab_Serializer).DeserializeObject<PlayFabJsonSuccess<<%- apiCall.result %>>>(resultRawJson);
+            var result = resultData.data;
+<%- getResultActions("            ", apiCall, api, true) %>
+            return new PlayFabResult<<%- apiCall.result %>> { Result = result, CustomData = customData };
+        } <% if (apiCall.name == "WriteTelemetryEvents") {
```
Comment (rgomez391) [inline targets/csharp/templates/PlayFab_Events_InstanceAPI.cs.ejs:123]: Only doing this for WriteTelemetryEvents API. Basically, overloading the API with a telemetry key string param. The existing endpoints will remain the same.

File: targets/csharp/templates/PlayFab_Events_API.cs.ejs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                PlayFabSettings.GlobalErrorHandler?.Invoke(error);
+                return new PlayFabResult<<%- apiCall.result %>> { Error = error, CustomData = customData };
+            }
+
+            var resultRawJson = (string)httpResult;
+            var resultData = PluginManager.GetPlugin<ISerializerPlugin>(PluginContract.PlayFab_Serializer).DeserializeObject<PlayFabJsonSuccess<<%- apiCall.result %>>>(resultRawJson);
+            var result = resultData.data;
+<%- getResultActions("            ", apiCall, api, false) %>
+            return new PlayFabResult<<%- apiCall.result %>> { Result = result, CustomData = customData };
+        } <% if (apiCall.name == "WriteTelemetryEvents") {
```
Comment (rgomez391) [inline targets/csharp/templates/PlayFab_Events_API.cs.ejs:66]: Only doing this for WriteTelemetryEvents API. Basically, overloading the API with a telemetry key string param. The existing endpoints will remain the same.

**Comment by rgomez391** [comment]
This is how the API will end up looking:

```csharp
#if !DISABLE_PLAYFABENTITY_API && !DISABLE_PLAYFAB_STATIC_API

using PlayFab.EventsModels;
using PlayFab.Internal;
#pragma warning disable 0649
using System;
// This is required for the Obsolete Attribute flag
//  which is not always present in all API's
#pragma warning restore 0649
using System.Collections.Generic;
using System.Threading.Tasks;

namespace PlayFab
{
    /// <summary>
    /// Write custom PlayStream and Telemetry events for any PlayFab entity. Telemetry events can be used for analytic,
    /// reporting, or debugging. PlayStream events can do all of that and also trigger custom actions in near real-time.
    /// </summary>
    public static class PlayFabEventsAPI
    {
        /// <summary>
        /// Verify entity login.
        /// </summary>
        public static bool IsEntityLoggedIn()
        {
            return PlayFabSettings.staticPlayer.IsEntityLoggedIn();
        }

        /// <summary>
        /// Clear the Client SessionToken which allows this Client to call API calls requiring login.
        /// A new/fresh login will be required after calling this.
        /// </summary>
        public static void ForgetAllCredentials()
        {
            PlayFabSettings.staticPlayer.ForgetAllCredentials();
        }

        /// <summary>
        /// Write batches of entity based events to PlayStream. The namespace of the Event must be 'custom' or start with 'custom.'.
        /// </summary>
        public static async Task<PlayFabResult<WriteEventsResponse>> WriteEventsAsync(WriteEventsRequest request, object customData = null, Dictionary<string, string> extraHeaders = null)
        {
            await new PlayFabUtil.SynchronizationContextRemover();

            var requestContext = request?.AuthenticationContext ?? PlayFabSettings.staticPlayer;
            var requestSettings = PlayFabSettings.staticSettings;
            if (requestContext.EntityToken == null) throw new PlayFabException(PlayFabExceptionCode.EntityTokenNotSet, "Must call Client Login or GetEntityToken before calling this method");


            var httpResult = await PlayFabHttp.DoPost("/Event/WriteEvents", request, "X-EntityToken", requestContext.EntityToken, extraHeaders);
            if (httpResult is PlayFabError)
            {
                var error = (PlayFabError)httpResult;
                PlayFabSettings.GlobalErrorHandler?.Invoke(error);
                return new PlayFabResult<WriteEventsResponse> { Error = error, CustomData = customData };
            }

            var resultRawJson = (string)httpResult;
            var resultData = PluginManager.GetPlugin<ISerializerPlugin>(PluginContract.PlayFab_Serializer).DeserializeObject<PlayFabJsonSuccess<WriteEventsResponse>>(resultRawJson);
            var result = resultData.data;

            return new PlayFabResult<WriteEventsResponse> { Result = result, CustomData = customData };
        } 

        /// <summary>
        /// Write batches of entity based events to as Telemetry events (bypass PlayStream). The namespace must be 'custom' or start
        /// with 'custom.'
        /// </summary>
        public static async Task<PlayFabResult<WriteEventsResponse>> WriteTelemetryEventsAsync(WriteEventsRequest request, object customData = null, Dictionary<string, string> extraHeaders = null)
        {
            await new PlayFabUtil.SynchronizationContextRemover();

            var requestContext = request?.AuthenticationContext ?? PlayFabSettings.staticPlayer;
            var requestSettings = PlayFabSettings.staticSettings;
            if (requestContext.EntityToken == null) throw new PlayFabException(PlayFabExceptionCode.EntityTokenNotSet, "Must call Client Login or GetEntityToken before calling this method");


            var httpResult = await PlayFabHttp.DoPost("/Event/WriteTelemetryEvents", request, "X-EntityToken", requestContext.EntityToken, extraHeaders);
            if (httpResult is PlayFabError)
            {
                var error = (PlayFabError)httpResult;
                PlayFabSettings.GlobalErrorHandler?.Invoke(error);
                return new PlayFabResult<WriteEventsResponse> { Error = error, CustomData = customData };
            }

            var resultRawJson = (string)httpResult;
            var resultData = PluginManager.GetPlugin<ISerializerPlugin>(PluginContract.PlayFab_Serializer).DeserializeObject<PlayFabJsonSuccess<WriteEventsResponse>>(resultRawJson);
            var result = resultData.data;

            return new PlayFabResult<WriteEventsResponse> { Result = result, CustomData = customData };
        } 

        /// <summary>
        /// Write batches of entity based events to as Telemetry events (bypass PlayStream) using a Telemetry Key. The namespace must be 'custom' or start
        /// with 'custom.'
        /// </summary>
        public static async Task<PlayFabResult<WriteEventsResponse>> WriteTelemetryEventsAsync(WriteEventsRequest request, string telemetryKey, object customData = null, Dictionary<string, string> extraHeaders = null)
        {
            await new PlayFabUtil.SynchronizationContextRemover();

            var httpResult = await PlayFabHttp.DoPost("/Event/WriteTelemetryEvents", request, "X-TelemetryKey", telemetryKey, extraHeaders);
            if (httpResult is PlayFabError)
            {
                var error = (PlayFabError)httpResult;
                PlayFabSettings.GlobalErrorHandler?.Invoke(error);
                return new PlayFabResult<WriteEventsResponse> { Error = error, CustomData = customData };
            }

            var resultRawJson = (string)httpResult;
            var resultData = PluginManager.GetPlugin<ISerializerPlugin>(PluginContract.PlayFab_Serializer).DeserializeObject<PlayFabJsonSuccess<WriteEventsResponse>>(resultRawJson);
            var result = resultData.data;

            return new PlayFabResult<WriteEventsResponse> { Result = result, CustomData = customData };
        } 
    }
}
#endif
```

**Review by jasonsandlin — APPROVED** [review]

**Comment by AlonsoMondal** [comment]
C# Pipeline run with these changes taken into account running green:
https://dev.azure.com/PlayFabInternal/Main/_build/results?buildId=187368&view=results
PR to preview changes made (should not be merged):
https://github.com/PlayFab/CSharpSDK/pull/122

**Pull request merged**